sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/OrderedListUserState.java

```diff
@@ -0,0 +1,314 @@
+    // accessible in pre-existing iterables even after:
+    //   (1) a sort key is added to or removed from pendingAdds, or
+    //   (2) a new value is added to an existing sort key
+    ArrayList<PrefetchableIterable<TimestampedValue<T>>> pendingAddsInRange = new ArrayList<>();
+    for (Entry<Instant, Collection<T>> kv :
+        pendingAdds.subMap(minTimestamp, limitTimestamp).entrySet()) {
+      pendingAddsInRange.add(
+          PrefetchableIterables.limit(
+              Iterables.transform(kv.getValue(), (v) -> TimestampedValue.of(v, kv.getKey())),
+              kv.getValue().size()));
+    }
+    Iterable<TimestampedValue<T>> valuesInRange = Iterables.concat(pendingAddsInRange);
+
+    if (!isCleared) {
+      StateRequest.Builder getRequestBuilder = this.request.toBuilder();
+      getRequestBuilder
+          .getStateKeyBuilder()
+          .getOrderedListUserStateBuilder()
+          .getRangeBuilder()
+          .setStart(minTimestamp.getMillis())
+          .setEnd(limitTimestamp.getMillis());
+
+      // TODO: consider use cache here
+      CachingStateIterable<TimestampedValue<T>> persistentValues =
+          StateFetchingIterators.readAllAndDecodeStartingFrom(
+              Caches.noop(),
+              this.beamFnStateClient,
+              getRequestBuilder.build(),
+              this.timestampedValueCoder);
+
+      // Make a snapshot of the current pendingRemoves and use them to filter persistent values.
+      // The values of pendingRemoves are kept, so that they will still be accessible in
+      // pre-existing iterables even after a sort key is removed.
+      TreeRangeSet<Instant> pendingRemovesSnapshot = TreeRangeSet.create(pendingRemoves);
+      Iterable<TimestampedValue<T>> persistentValuesAfterRemoval =
+          Iterables.filter(
+              persistentValues, v -> !pendingRemovesSnapshot.contains(v.getTimestamp()));
+
+      return Iterables.mergeSorted(
+          ImmutableList.of(persistentValuesAfterRemoval, valuesInRange),
+          Comparator.comparing(TimestampedValue::getTimestamp));
+    }
+
+    return valuesInRange;
+  }
+
+  public Iterable<TimestampedValue<T>> read() {
+    checkState(
+        !isClosed,
+        "OrderedList user state is no longer usable because it is closed for %s",
+        request.getStateKey());
+
+    return readRange(Instant.ofEpochMilli(Long.MIN_VALUE), Instant.ofEpochMilli(Long.MAX_VALUE));
+  }
+
+  public void clearRange(Instant minTimestamp, Instant limitTimestamp) {
+    checkState(
+        !isClosed,
+        "OrderedList user state is no longer usable because it is closed for %s",
+        request.getStateKey());
+
+    // Remove items (in a collection) in the specific range from pendingAdds.
+    // The old values of the removed sub map are kept, so that they will still be accessible in
+    // pre-existing iterables even after the sort key is cleared.
+    pendingAdds.subMap(minTimestamp, limitTimestamp).clear();
+    if (!isCleared) {
```

Comment:
Question about sequence of events. Suppose we call `clear()`, then add some elements, then call `clearRange()`. I'm worried we'll then return those elements if we call `readRange()` because they won't be in `pendingRemoves`. I suspect we want to get rid of this `if` and always add to `pendingRemoves`.

Reply:
(If that's the case, make sure this is covered by tests.)

Reply:
I think there may be some confusion of what pendingRemoves actually is. In this implementation, it contains the items we need to remove **from the persisted storage.** 

Let's look at your use case. After we call clear(), all pendingAdds and pendingRemoves are reset to empty, which means that we expect the persisted storage to restart from scratch. Then if you call add() and then clearRange(), the operations will only be done on pendingAdds() (i.e. we do a subMap() on pendingAdds on Line 248). 

We have the condition on Line 249, because we don't need to remove anything from that empty persisted storage.